src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,158 @@
+		/*if (s_handles.contains(path)) {
+			m_library = s_handles[path];
+			m_library->ref();
+		} else {*/
+			m_library = mLibraryLoad(path.toUtf8().constData());
+			/*if (m_library != nullptr) {
+				//s_handles[path] = m_library; // FIXME!
+			} else {
+				m_library = mLibraryCreateEmpty();
+			}
+		}*/
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	QList<mLibraryEntry*> newItems;
+	mLibraryListing m_listings;
+	mLibraryListingInit(&m_listings, 0);
+	mLibraryGetEntries(m_library, &m_listings, 0, 0, /*m_constraints.base ? &m_constraints :*/ nullptr);
+	for (size_t i = 0; i < mLibraryListingSize(&m_listings); i++)
+		newItems.append(mLibraryListingGetPointer(&m_listings, i));
+
+	m_tree_widget = new LibraryTree(this);
+	addWidget(m_tree_widget);
+	connect(m_tree_widget, &LibraryTree::StartGame, this, &LibraryController::startGame);
+
+	m_grid_widget = new LibraryGrid(this);
+	addWidget(m_grid_widget);
+	connect(m_grid_widget, &LibraryGrid::StartGame, this, &LibraryController::startGame);
+
+	m_tree_widget->AddGames(newItems);
+	m_grid_widget->AddGames(newItems);
+
+	SetViewStyle(GameListStyle::STYLE_LIST);
+}
+
+LibraryController::~LibraryController()
+{
+	for (mLibraryEntry* file : m_games.values())
+		delete file;
+}
+
+void LibraryController::SetViewStyle(GameListStyle newStyle)
+{
+	if (newStyle == m_current_style)
+		return;
+	m_current_style = newStyle;
+
+	if (newStyle == GameListStyle::STYLE_LIST || newStyle == GameListStyle::STYLE_TREE)	{
+		m_tree_widget->SelectGame(selectedGame());
+		setCurrentWidget(m_tree_widget);
+		m_tree_widget->SetViewStyle(newStyle);
+	} else {
+		m_grid_widget->SelectGame(selectedGame());
+		setCurrentWidget(m_grid_widget);
+		m_grid_widget->SetViewStyle(newStyle);
+	}
+}
+
+mLibraryEntry* LibraryController::selectedGame()
+{
+	if (currentWidget() == m_grid_widget)
+		return m_grid_widget->SelectedGame();
+	else
+		return m_tree_widget->SelectedGame();
+}
+
+VFile* LibraryController::selectedVFile()
+{
+	mLibraryEntry* selectedEntry = selectedGame();
+	if (selectedEntry)
+		return mLibraryOpenVFile(m_library, selectedEntry);
+	else
+		return nullptr;
+}
+
+void LibraryController::addDirectory(const QString& dir)
+{
+	setDisabled(true);
+
+	//mLibraryLoadDirectory(m_library, dir.toUtf8().constData());
+
+	/* TODO!
+	QList<mLibraryEntry*> newItems;
+	QMap<QString, mLibraryEntry*> allItems;
+
+	// Check for games that were removed
+	QList<mLibraryEntry*> removedGames;
+	for (QString& path : m_games.keys()) {
+		if (!allItems.contains(path)) {
+			removedGames.append(m_games.value(path));
+			m_games.remove(path);
+		}
+	}
+
+	m_tree_widget->AddGames(newItems);
+	m_grid_widget->AddGames(newItems);
+
+	m_tree_widget->RemoveGames(removedGames);
+	m_grid_widget->RemoveGames(removedGames);
+
+	for (mLibraryEntry* file : removedGames)
+		delete file;
+		*/
+
+	setDisabled(false);
+}
+
+void LibraryController::selectLastBootedGame()
```

Comment:
Dolphin had this feature where on startup it would select the last booted game in the gamelist. Implement or remove?